### PR TITLE
Add test for yast2 rdp service

### PR DESCRIPTION
**Pull request by badboywj**
Add yast-rdp.pm.
yast-rdp.pm:   this module will test yast rdp service in cmd mode.

### CODE REVIEW

File: yast2_rdp.pm [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+# Maintainer: Jun Wang <jgwang@suse.com>
+
+use base "y2_module_guitest";
+use strict;
+use warnings;
+use testapi;
+use utils;
+use version_utils 'is_sle';
+
+my $repos = {'12-SP1' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP1/SLE-12-SP1/',
```
Comment (mimi1vx) [inline yast2_rdp.pm]: this should not be In any test, if isn't yast2-rdp in the product simply not run on this product

File: yast2_rdp.pm [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    '12-SP1-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP1/SLE-12-SP1_Update/',
+    '12-SP2'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP2/SLE_12_SP2/',
+    '12-SP2-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP2/SUSE_SLE-12-SP2_Update/',
+    '12-SP3'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP3/SLE_12_SP3/',
+    '12-SP3-UP' => '',
+    '12-SP4'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP4/SLE_12_SP4/',
+    '12-SP4-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP4/SUSE_SLE-12-SP4_Update/'
+};
+
+sub becomeroot {
```
Comment (mimi1vx) [inline yast2_rdp.pm]: ? , what about `select_console` or http://open.qa/api/testapi/#_become_root ?

File: yast2_rdp.pm [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        if ("$version" ne '12-SP3') {
+            zypper_ar("$repos->{qq($version-UP)}", 'Devel-YaST-SLE-UP');
+        }
+    } elsif (is_sle('<12-SP3')) {
+        die "This version of product is NOT supported.";
+    }
+    zypper_call("in yast2-rdp", exitcode => [0, 102, 103]);
+    type_string "exit\n";
+
+    # disable screen saver.
```
Comment (mimi1vx) [inline yast2_rdp.pm]: why ?

File: yast_rdp.pm [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+use version_utils 'is_sle';
+
+my $repos = {'12-SP1' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP1/SLE-12-SP1/',
+    '12-SP1-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP1/SLE-12-SP1_Update/',
+    '12-SP2'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP2/SLE_12_SP2/',
+    '12-SP2-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP2/SUSE_SLE-12-SP2_Update/',
+    '12-SP3'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP3/SLE_12_SP3/',
+    '12-SP3-UP' => '',
+    '12-SP4'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP4/SLE_12_SP4/',
+    '12-SP4-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP4/SUSE_SLE-12-SP4_Update/'
```
Comment (plusky) [inline yast_rdp.pm]: I am kinda curious which product is being tested, or to rephrease my question, is tested scenario something that is supported?

File: yast2_rdp.pm [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+my $repos = {'12-SP1' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP1/SLE-12-SP1/',
+    '12-SP1-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP1/SLE-12-SP1_Update/',
+    '12-SP2'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP2/SLE_12_SP2/',
+    '12-SP2-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP2/SUSE_SLE-12-SP2_Update/',
+    '12-SP3'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP3/SLE_12_SP3/',
+    '12-SP3-UP' => '',
+    '12-SP4'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP4/SLE_12_SP4/',
+    '12-SP4-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP4/SUSE_SLE-12-SP4_Update/'
+};
```
Comment (czerw) [inline yast2_rdp.pm]: Why are you adding development repositories ? If it is available only for SLE15+ in default repos, it shoudl be tested only for SLE15+.

File: yast2_rdp.pm [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    if (!($start and $enable)) {
+        die "It failed that yast2-rdp starts xrdp service";
+    }
+
+    # yast2-rdp stops xrdp service.
+    $self->launch_yast2_module_x11("rdp", match_timeout => 120);
+    check_box('RDP', 'yast2-rdp-ui-enable-xrdp');
+
+    # check if xrdp service stops.
+    $start  = check_service('General', 'echo $(systemctl is-active xrdp.service)');
```
Comment (mimi1vx) [inline yast2_rdp.pm]: `systemctl` from utils.pm .....

File: yast_rdp.pm [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    script_run("systemctl disable xrdp.service", quiet => 1);
+
+    # start xrdp service
+    script_run("yast2 rdp allow verbose set=yes", quiet => 1);
+
+    # check if xrdp service starts.
+    my $start  = check_service('echo $(systemctl is-active xrdp.service)');
+    my $enable = check_service('echo $(systemctl is-enabled xrdp.service)');
+    if (!($start and $enable)) {
+        die "It failed that yast2-rdp starts xrdp service";
```
Comment (plusky) [inline yast_rdp.pm]: I have to admit that this sentence does not make that much sense

File: yast2_rdp.pm [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+    # yast2-rdp deletes "ms-wbt" service from these zones "external home internal public trusted work"
+    # closing port in firewall.
+    $self->launch_yast2_module_x11("rdp", match_timeout => 120);
+    check_box('Firewall', 'yast2-rdp-ui-open-port-in-firewall');
+
+    # check if closing port in firewall correctly.
+    $start = check_service('General', 'echo $(systemctl is-active firewalld.service)');
+    if ($start) {
+        my $service = check_service('Firewall', 'echo $(firewall-cmd --list-services | grep -Eo "ms-wbt")X');
```
Comment (mimi1vx) [inline yast2_rdp.pm]: on SLE12 is SuSEFirewall2 ....

File: yast2_rdp.pm [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        }
+    }
+
+    # yast2-rdp deletes "ms-wbt" service from these zones "external home internal public trusted work"
+    # closing port in firewall.
+    $self->launch_yast2_module_x11("rdp", match_timeout => 120);
+    check_box('Firewall', 'yast2-rdp-ui-open-port-in-firewall');
+
+    # check if closing port in firewall correctly.
+    $start = check_service('General', 'echo $(systemctl is-active firewalld.service)');
```
Comment (mimi1vx) [inline yast2_rdp.pm]: on SLE12 is SuSEFirewall2 .... , on SLE15+ , and Leap15+ firewalld

File: yast_rdp.pm [hunk 10]
```diff
@@ -0,0 +1,90 @@
+# SUSE's openQA tests
```
Comment (mimi1vx) [inline yast_rdp.pm]: this file is for ??

File: yast2_rdp.pm (hunk 2 above)
Comment (czerw) [inline yast2_rdp.pm]: there  is already existing function become_root, why are you creating new one?

File: yast_rdp.pm [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        zypper_ar("$repos->{$version}", 'Devel-YaST-SLE');
+        if ("$version" ne '12-SP3') {
+            zypper_ar("$repos->{qq($version-UP)}", 'Devel-YaST-SLE-UP');
+        }
+    } elsif (is_sle('<12-SP3')) {
+        die "This version of product is NOT supported.";
+    }
+
+    zypper_call("in yast2-rdp", exitcode => [0, 102, 103]);
+    script_run("systemctl stop xrdp.service",    quiet => 1);
```
Comment (plusky) [inline yast_rdp.pm]: I guess there is some sort of wrappers for enabling/disabling services ...

**Comment by czerw** [comment]
> yast2-rdp.pm: this module will test yast2 rdp service with gui.
> yast-rdp.pm: this module will test yast rdp service in cmd mode.

gui test should be in directory tests/yast2_gui
command line test should be in directory  tests/yast2_cmd

**Comment by czerw** [comment]
Please provide verification runs, also tests are not scheduled in main.pm.

File: yast2_rdp.pm [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+    wait_still_screen 2;
+
+    if ($type eq 'General') {
+        if ($result eq "active" or $result eq "enabled") {
+            return 1;
+        } else {
+            return 0;
+        }
+    } elsif ($type eq 'Firewall') {
+        if ($result eq 'ms-wbtX') {
```
Comment (czerw) [inline yast2_rdp.pm]: what is `ms-wbtX`?

**Comment by plusky** [comment]
There also seeems to be some tests for RDP in tests/x11/remote_desktop

File: yast_rdp.pm (hunk 11 above)
Comment (okurz) [reply yast_rdp.pm]: There is, call `systemctl 'stop xrdp'`

File: yast_rdp.pm [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    # yast2-rdp was added after SLE12SP2.
+    if (is_sle('>12-SP2') and is_sle('<15')) {
+        my $version = get_var('VERSION');
+        zypper_ar("$repos->{$version}", 'Devel-YaST-SLE');
+        if ("$version" ne '12-SP3') {
+            zypper_ar("$repos->{qq($version-UP)}", 'Devel-YaST-SLE-UP');
+        }
+    } elsif (is_sle('<12-SP3')) {
+        die "This version of product is NOT supported.";
+    }
```
Comment (okurz) [inline yast_rdp.pm]: The include/exclude of test modules based on product versions is normally handled in the test schedule, better take a look in `lib/main_common.pm` and `products/opensuse/main.pm` and `products/sle/main.pm`

File: yast_rdp.pm [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    zypper_call("in yast2-rdp", exitcode => [0, 102, 103]);
+    script_run("systemctl stop xrdp.service",    quiet => 1);
+    script_run("systemctl disable xrdp.service", quiet => 1);
+
+    # start xrdp service
+    script_run("yast2 rdp allow verbose set=yes", quiet => 1);
+
+    # check if xrdp service starts.
+    my $start  = check_service('echo $(systemctl is-active xrdp.service)');
+    my $enable = check_service('echo $(systemctl is-enabled xrdp.service)');
```
Comment (okurz) [inline yast_rdp.pm]: I suggest to simply use `systemctl 'is-active xrdp';` and `systemctl 'is-enabled xrdp';` which will also check that the services are actually started and enabled

File: yast_rdp.pm [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    } elsif (is_sle('<12-SP3')) {
+        die "This version of product is NOT supported.";
+    }
+
+    zypper_call("in yast2-rdp", exitcode => [0, 102, 103]);
+    script_run("systemctl stop xrdp.service",    quiet => 1);
+    script_run("systemctl disable xrdp.service", quiet => 1);
+
+    # start xrdp service
+    script_run("yast2 rdp allow verbose set=yes", quiet => 1);
```
Comment (okurz) [inline yast_rdp.pm]: Better use `assert_script_run` to abort the test when this fails

File: yast_rdp.pm (hunk 4 above)
Comment (badboywj) [reply yast_rdp.pm]: I just had a try to support more products. If we only test yast2-rdp in the default repos, I can adjust the code.

File: yast2_rdp.pm (hunk 5 above)
Comment (badboywj) [reply yast2_rdp.pm]: If we only test yast2-rdp in the default repos, I can adjust the code.

File: yast_rdp.pm (hunk 7 above)
Comment (badboywj) [reply yast_rdp.pm]: If I am checking if yast2-rdp works well, I don't think that we start yast2-rdp to check itself, it is better that using another tool check it. So I add this code. Of course, I think the best way is that using "assert_screen" check if yast2-rdp works well.

File: yast2_rdp.pm (hunk 1 above)
Comment (badboywj) [reply yast2_rdp.pm]: yes, if we only test yast2-rdp in the default repos, I will adjust the code.

File: yast2_rdp.pm (hunk 2 above)
Comment (badboywj) [reply yast2_rdp.pm]: Actually I had a try "become_root" testapi, but it produced too many reference pictures, this is not good to review the result. So I simplified "become_root" to "becomeroot" for reducing reference pictures.

File: yast2_rdp.pm (hunk 3 above)
Comment (badboywj) [reply yast2_rdp.pm]: you know, when runing yast2-rdp.pm, this packages maybe not be installed, this makes sure yast2-rdp package is installed, of course, if it was installed, yast2-rdp.pm will continue.

File: yast2_rdp.pm (hunk 6 above)
Comment (badboywj) [reply yast2_rdp.pm]: Thank you for your advice, I will have a try.

File: yast_rdp.pm (hunk 11 above)
Comment (badboywj) [reply yast_rdp.pm]: Thank you guys, it does. I will have a try.

File: yast2_rdp.pm (hunk 8 above)
Comment (badboywj) [reply yast2_rdp.pm]: yes, thanks for your advice.

File: yast2_rdp.pm (hunk 9 above)
Comment (badboywj) [reply yast2_rdp.pm]: yes, thanks for your advice.

File: yast_rdp.pm (hunk 13 above)
Comment (badboywj) [reply yast_rdp.pm]: Yes, I see. Thank you.

File: yast_rdp.pm (hunk 14 above)
Comment (badboywj) [reply yast_rdp.pm]: OK.

File: yast_rdp.pm (hunk 15 above)
Comment (badboywj) [reply yast_rdp.pm]: OK.

File: yast2_rdp.pm (hunk 12 above)
Comment (badboywj) [reply yast2_rdp.pm]: ms-wbt is a return value, it means that the port of xrdp is open in firewall. X, it is just a habit for fear that "$result" is null

File: yast_rdp.pm (hunk 7 above)
Comment (okurz) [reply yast_rdp.pm]: the comment was about the die-message. Maybe you want to write "yast2-rdp failed to start the xrdp service"?

File: yast2_rdp.pm (hunk 2 above)
Comment (okurz) [reply yast2_rdp.pm]: Well, you could improve the existing `become_root` function but defining another one that is similarly named and confusing is most certainly not the way to go

File: yast_rdp.pm (hunk 7 above)
Comment (badboywj) [reply yast_rdp.pm]: > the comment was about the die-message. Maybe you want to write "yast2-rdp failed to start the xrdp service"?

Yes, you are right.  Sorry, I had a mistake in comprehension with "plysky".

File: yast2_rdp.pm (hunk 2 above)
Comment (badboywj) [reply yast2_rdp.pm]: > Well, you could improve the existing `become_root` function but defining another one that is similarly named and confusing is most certainly not the way to go

Yes, this point is important, I will care it.

File: yast_rdp.pm (hunk 10 above)
Comment (badboywj) [reply yast_rdp.pm]: you know, for yast2-rdp, there are two modes, one with gui, and the other is cmdline mode, like this "yast2 rdp allow set=yes" to start xrdp service. yast_rdp.pm is for testing in cmdline mode.

File: yast2_rdp.pm (hunk 2 above)
Comment (badboywj) [reply yast2_rdp.pm]: > there is already existing function become_root, why are you creating new one?

Actually I had a try "become_root" testapi, but it produced too many reference pictures, this is not good to review the result. So I simplified "become_root" to "becomeroot" for reducing reference pictures.

Comment (badboywj) [reply yast2_rdp.pm]: > there is already existing function become_root, why are you creating new one?

Actually I had a try "become_root" testapi, but it produced too many reference pictures, this is not good to review the result. So I simplified "become_root" to "becomeroot" for reducing reference pictures.

File: yast_rdp.pm (hunk 7 above)
Comment (plusky) [reply yast_rdp.pm]: poor "plysky"

Comment (badboywj) [reply yast_rdp.pm]: sorry for my typo

**Pull request closed without merging**